File: src/backend/folha/src/model/folha_afastamento_model.py
from src import db
from src.model.view_pessoa_colaborador_model import ViewPessoaColaboradorModel
from src.model.folha_tipo_afastamento_model import FolhaTipoAfastamentoModel


class FolhaAfastamentoModel(db.Model):
    __tablename__ = 'folha_afastamento'

    id = db.Column(db.Integer, primary_key=True)
    data_inicio = db.Column(db.DateTime)
    data_fim = db.Column(db.DateTime)
    dias_afastado = db.Column(db.Integer)
    id_colaborador = db.Column(db.Integer, db.ForeignKey('view_pessoa_colaborador.id'))
    id_folha_tipo_afastamento = db.Column(db.Integer, db.ForeignKey('folha_tipo_afastamento.id'))

    view_pessoa_colaborador_model = db.relationship('ViewPessoaColaboradorModel', foreign_keys=[id_colaborador])
    folha_tipo_afastamento_model = db.relationship('FolhaTipoAfastamentoModel', foreign_keys=[id_folha_tipo_afastamento])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_colaborador = data.get('idColaborador')
        self.id_folha_tipo_afastamento = data.get('idFolhaTipoAfastamento')
        self.data_inicio = data.get('dataInicio')
        self.data_fim = data.get('dataFim')
        self.dias_afastado = data.get('diasAfastado')

    def serialize(self):
        return {
            'id': self.id,
            'idColaborador': self.id_colaborador,
            'idFolhaTipoAfastamento': self.id_folha_tipo_afastamento,
            'dataInicio': self.data_inicio.isoformat(),
            'dataFim': self.data_fim.isoformat(),
            'diasAfastado': self.dias_afastado,
            'viewPessoaColaboradorModel': self.view_pessoa_colaborador_model.serialize() if self.view_pessoa_colaborador_model else None,
            'folhaTipoAfastamentoModel': self.folha_tipo_afastamento_model.serialize() if self.folha_tipo_afastamento_model else None,
        }

File: src/frontend/folha/lib/app/data/model/folha_afastamento_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:folha/app/infra/infra_imports.dart';
import 'package:folha/app/data/model/model_imports.dart';
import 'package:intl/intl.dart';

class FolhaAfastamentoModel {
	int? id;
	int? idColaborador;
	int? idFolhaTipoAfastamento;
	DateTime? dataInicio;
	DateTime? dataFim;
	int? diasAfastado;
	ViewPessoaColaboradorModel? viewPessoaColaboradorModel;
	FolhaTipoAfastamentoModel? folhaTipoAfastamentoModel;

	FolhaAfastamentoModel({
		this.id,
		this.idColaborador,
		this.idFolhaTipoAfastamento,
		this.dataInicio,
		this.dataFim,
		this.diasAfastado,
		this.viewPessoaColaboradorModel,
		this.folhaTipoAfastamentoModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'data_inicio',
		'data_fim',
		'dias_afastado',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Data Inicio',
		'Data Fim',
		'Dias Afastado',
	];

	FolhaAfastamentoModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idColaborador = jsonData['idColaborador'];
		idFolhaTipoAfastamento = jsonData['idFolhaTipoAfastamento'];
		dataInicio = jsonData['dataInicio'] != null ? DateTime.tryParse(jsonData['dataInicio']) : null;
		dataFim = jsonData['dataFim'] != null ? DateTime.tryParse(jsonData['dataFim']) : null;
		diasAfastado = jsonData['diasAfastado'];
		viewPessoaColaboradorModel = jsonData['viewPessoaColaboradorModel'] == null ? ViewPessoaColaboradorModel() : ViewPessoaColaboradorModel.fromJson(jsonData['viewPessoaColaboradorModel']);
		folhaTipoAfastamentoModel = jsonData['folhaTipoAfastamentoModel'] == null ? FolhaTipoAfastamentoModel() : FolhaTipoAfastamentoModel.fromJson(jsonData['folhaTipoAfastamentoModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idColaborador'] = idColaborador != 0 ? idColaborador : null;
		jsonData['idFolhaTipoAfastamento'] = idFolhaTipoAfastamento != 0 ? idFolhaTipoAfastamento : null;
		jsonData['dataInicio'] = dataInicio != null ? DateFormat('yyyy-MM-ddT00:00:00').format(dataInicio!) : null;
		jsonData['dataFim'] = dataFim != null ? DateFormat('yyyy-MM-ddT00:00:00').format(dataFim!) : null;
		jsonData['diasAfastado'] = diasAfastado;
		jsonData['viewPessoaColaboradorModel'] = viewPessoaColaboradorModel?.toJson;
		jsonData['folhaTipoAfastamentoModel'] = folhaTipoAfastamentoModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idColaborador = plutoRow.cells['idColaborador']?.value;
		idFolhaTipoAfastamento = plutoRow.cells['idFolhaTipoAfastamento']?.value;
		dataInicio = Util.stringToDate(plutoRow.cells['dataInicio']?.value);
		dataFim = Util.stringToDate(plutoRow.cells['dataFim']?.value);
		diasAfastado = plutoRow.cells['diasAfastado']?.value;
		viewPessoaColaboradorModel = ViewPessoaColaboradorModel();
		viewPessoaColaboradorModel?.nome = plutoRow.cells['viewPessoaColaboradorModel']?.value;
		folhaTipoAfastamentoModel = FolhaTipoAfastamentoModel();
		folhaTipoAfastamentoModel?.nome = plutoRow.cells['folhaTipoAfastamentoModel']?.value;
	}	

	FolhaAfastamentoModel clone() {
		return FolhaAfastamentoModel(
			id: id,
			idColaborador: idColaborador,
			idFolhaTipoAfastamento: idFolhaTipoAfastamento,
			dataInicio: dataInicio,
			dataFim: dataFim,
			diasAfastado: diasAfastado,
		);			
	}

	
}